### PR TITLE
Add docs and enable missing_docs lint in icu_provider

### LINKED ISSUE #2110: Enable warn(missing_docs) for all provider crates
All ICU provider crates should be fully documented. To track this, we should enable `warn(missing_docs)` as suggested in boilerplate.md.

### PR DESCRIPTION
**Pull request by sffc**
Fixes #2110

### CODE REVIEW

File: provider/core/src/error.rs [hunk 1]
```diff
@@ -263,3 +265,28 @@ impl From<std::io::Error> for DataError {
         DataErrorKind::Io(e.kind()).into_error()
     }
 }
+
+/// A [`DataError`] that also returns a [`DataPayload`] to the caller.
+#[allow(clippy::exhaustive_structs)] // this type is stable
+pub struct DataErrorWithPayload<M: DataMarker>(pub DataPayload<M>, pub DataError);
```
Comment (robertbastian) [inline provider/core/src/error.rs:271]: Why move this here? This is only used by the converter thing, which itself is only used for the zero-copy test. I don't think we should make this available outside `datagen`.

File: provider/core/src/marker.rs [hunk 2]
```diff
@@ -10,6 +10,8 @@ use crate::yoke::Yokeable;
 /// Trait marker for data structs. All types delivered by the data provider must be associated with
 /// something implementing this trait.
 ///
+/// Structs implementing this trait are normally generated with the [`data_struct!`] macro.
```
Comment (robertbastian) [inline provider/core/src/marker.rs:13]: Not a function-like macro.

```suggestion
/// Structs implementing this trait are normally generated with the [`#[data_struct]`] macro.
```

File: provider/core/src/request.rs [hunk 3]
```diff
@@ -28,13 +34,72 @@ impl fmt::Display for DataRequest<'_> {
...
+///
+/// let locale1 = "en-u-ca-buddhist".parse::<Locale>().unwrap();
+/// let data_locale: DataLocale = locale1.into();
+/// let locale2 = data_locale.into_locale();
+///
+/// assert_eq!(locale2.to_string(), "en-u-ca-buddhist");
+/// ```
+///
+/// You can alternatively create a [`DataLocale`] from a borrowed [`Locale`], which is more
+/// efficient than cloning the [`Locale`]:
```
Comment (robertbastian) [inline provider/core/src/request.rs:65]: ...but less efficient than converting an owned `Locale`

File: provider/core/src/request.rs [hunk 4]
```diff
@@ -28,13 +34,72 @@ impl fmt::Display for DataRequest<'_> {
...
+///
+/// ```
+/// use icu_locid::Locale;
+/// use icu_provider::DataLocale;
+///
+/// let locale1 = "en-u-ca-buddhist".parse::<Locale>().unwrap();
+/// let data_locale: DataLocale = (&locale1).into();
```
Comment (robertbastian) [inline provider/core/src/request.rs:72]: Nit: I'd prefer writing this as
```rust
let data_locale = DataLocale::from(&locale);
```
which is a bit shorter and feels more explicit as well. `(&locale1).into()` requires the type information from outside the expression, which is cool if it's there, but not cool if you have to explicitly type the variable. Also in the other examples.

File: provider/core/src/serde/mod.rs [hunk 5]
```diff
@@ -83,6 +86,30 @@ where
 }
 
 impl DataPayload<BufferMarker> {
+    /// Deserialize a [`DataPayload`]`<`[`BufferMarker`]`>` into a [`DataPayload`] of a
+    /// specific concrete type.
+    ///
+    /// The buffer format must be known externally to this function; it is normally carried
```
Comment (robertbastian) [inline provider/core/src/serde/mod.rs:92]: Nit: awkward phrasing. I also don't see the value of this statement. Yes the buffer format is an input, but we can already see that. That it's in the DataResponseMetadata is an implementation detail of as_deserializing.

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: provider/core/src/error.rs (hunk 1 above)
Comment (sffc) [reply provider/core/src/error.rs:271]: It seems like a generally useful type; if we added some function that needed to use this type outside of `icu_provider::datagen` (which could be either inside or outside of `icu_provider`), we would need to either add a similar duplicated type, point into `icu_provider::datagen` to get it at that point, or move the type and re-export it from `icu_provider::datagen`.

File: provider/core/src/marker.rs (hunk 2 above)
Comment (Manishearth) [reply provider/core/src/marker.rs:13]: that linkification may not work, try `[....](macro@data_struct)`

File: provider/core/src/request.rs (hunk 4 above)
Comment (Manishearth) [reply provider/core/src/request.rs:72]: +1, we should prefer from over `: DataLocale`

File: provider/core/src/response.rs [hunk 6]
```diff
@@ -570,6 +570,28 @@ where
         })
     }
 
+    /// Convert between two [`DataMarker`] types that are compatible with each other.
+    ///
+    /// This happens if they both have the same [`DataMarker::Yokeable`] type.
+    ///
+    /// Can be used to erase the key of a data payload in cases where multiple keys correspond
```
Comment (Manishearth) [inline provider/core/src/response.rs:577]: nit:

```suggestion
    /// Can be used to type-erase the key of a data payload in cases where multiple keys correspond
```

**Review by Manishearth — APPROVED** [review]

File: provider/core/src/response.rs (hunk 6 above)
Comment (sffc) [reply provider/core/src/response.rs:577]: Mmm, we're not erasing the type. We are actually erasing the key, by downgrading from a KeyedDataMarker to a DataMarker.

File: provider/core/src/error.rs (hunk 1 above)
Comment (robertbastian) [reply provider/core/src/error.rs:271]: > ... or move the type and re-export it from icu_provider::datagen.

That's a non-breaking change which we can do if we ever need that. I'm not convinced that it's a generally useful type, I find it quite confusing actually.

File: provider/core/src/request.rs (hunk 4 above)
Comment (robertbastian) [reply provider/core/src/request.rs]: The other two examples as well

**Review by robertbastian — COMMENTED** [review]
.

**Review by Manishearth — APPROVED** [review]

File: provider/core/src/error.rs (hunk 1 above)
Comment (sffc) [reply provider/core/src/error.rs:271]: Okay, it seems you feel more strongly on this than I. I'll move it back to datagen; do you have an opinion on the name `DataErrorWithPayload` versus the original name `ReturnedPayloadError`? (I prefer `DataErrorWithPayload`)

Comment (robertbastian) [reply provider/core/src/error.rs:271]: No strong opinion, I'd also be happy to not name this and just use a pair.

Comment (sffc) [reply provider/core/src/error.rs:271]: I think it has a name so we can implement the `?` operator for it

But I'd be okay changing the small number of call sites to use `map_err`. Seems like that may be the easiest solution.

Comment (robertbastian) [reply provider/core/src/error.rs:271]: You should be able to define `From<(DataError, DataPayload<M>)> for DataError`. But I'd prefer using `map_err` over this.

Comment (sffc) [reply provider/core/src/error.rs:271]: Removed. Turns out the `?` appears to never be used anywhere on this error?

File: provider/core/src/request.rs (hunk 4 above)
Comment (sffc) [reply provider/core/src/request.rs]: done

**Review by Manishearth — APPROVED** [review]

**Review by robertbastian — APPROVED** [review]

**Pull request merged**